docs/basics/101-133-containersrun.rst

```diff
@@ -0,0 +1,306 @@
+   `install <https://singularity.lbl.gov/docs-installation>`_ the software singularity.
+
+Using ``datalad containers``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. index:: ! datalad command; containers-add
+.. index:: ! datalad command; containers-run
+
+One core feature of the ``datalad containers`` extension is that it registers
+computational containers to a dataset. This is done with the
+:command:`datalad containers-add` command.
+Once a container is registered, arbitrary commands can be executed inside of
+it, i.e., in the precise software environment the container encapsulates. All it
+needs for this it to swap the :command:`datalad run` command introduced in
+section :ref:`run` with the :command:`datalad containers-run` command.
+
+Let's see this in action for the ``midterm_analysis`` dataset by rerunning
+the analysis you did for the midterm project within a Singularity container.
+We start by registering a container to the dataset.
+For this we will pull a container from Singularity hub. This container has
+been build for the handbook, and it contains the relevant Python setup for
+the analysis. If you're curious how to create a Singularity image, the hidden
+section below has some pointers:
+
+.. findoutmore:: How to make a Singularity image
+
+   Singularity containers are build from Image files, often
+   called "recipes", that hold a "definition" of the software container and its
+   contents and components. The
+   `singularity documentation <https://sylabs.io/guides/3.4/user-guide/build_a_container.html>`_
+   has its own tutorial on how to build such Images from scratch.
+   An alternative to writing the Image file by hand is to use
+   `Neurodocker <https://github.com/kaczmarj/neurodocker#singularity>`_. This
+   command-line program can help you generate custom Singularity recipes (and
+   also ``Dockerfiles``, from which Docker images are build). A wonderful tutorial
+   on how to use Neurodocker is
+   `this introduction <https://miykael.github.io/nipype_tutorial/notebooks/introduction_neurodocker.html>`_
+   by Michael Notter.
+
+   Once a recipe exists, the command
+
+   .. code-block:: bash
+
+      sudo singularity build <NAME> <RECIPE>
+
+   will build a container (called ``<NAME>``) from the recipe. Alternatively,
+   `Singularity Hub <https://singularity-hub.org/>`_ integrates with Github
+   and builds containers from Images pushed to repositories on Github.
+   `The docs <https://singularityhub.github.io/singularityhub-docs/>`_
+   give you a set of instructions on how to do this.
+
+The :command:`datalad containers-add` command takes an arbitrary
+name to give to the container, and a path or url to a container image:
+
+.. runrecord:: _examples/DL-101-133-101
+   :language: console
+   :workdir: dl-101/DataLad-101/midterm_project
+
+   # we are in the midterm_project subdataset
+   $ datalad containers-add python --url shub://adswa/resources:1
```

Comment:
I would likely choose a different, less generic name. It might be confusing to need to have Python installed on a system, in order to run Datalad and `datalad containers-add` to ultimately have a an image with Python that we could not execute with `datalad containers-run` without also having Python outside of the container ...

Reply:
Maybe also worth pointing out the source of that image: https://github.com/datalad-handbook/resources

Reply:
that's a very good point. Do you have a preference for any of the following, or other ideas?

- ``iris_env``
- ``midterm-software``